devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy

```diff
@@ -327,35 +311,79 @@ private def makeSerialResult(serial) {
         value: serial,
         descriptionText: "${linkText} has serial ${serial}" ]
 }
-/**** COMMAND METHODS ****/
-// def mfgCode() {
-// 	 ["zcl mfg-code 0x115B", "delay 200"]
-// }
 
+// takes a level from 0 to 100 and translates it to a ZigBee move to level with on/off command
+private def makeLevelCommand(level) {
+    def rangeMax = 254
+    def scaledLevel = Math.round(level * rangeMax / 100)
+    log.debug "scaled level for ${level}%: ${scaledLevel}"
+
+    // convert to hex string and pad to two digits
+    def hexLevel = new BigInteger(scaledLevel.toString()).toString(16).padLeft(2, '0')
+
+    "st cmd 0x${device.deviceNetworkId} 1 8 4 {${hexLevel} 0000}"
+}
+
+/**** COMMAND METHODS ****/
 def on() {
-    log.debug "on()"
+    def linkText = getLinkText(device)
+    log.debug "open ${linkText}"
+
+    // only change the state if the vent is not obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot open because ${linkText} is obstructed")
+        return
+    }
+
     sendEvent(makeOnOffResult(1))
     "st cmd 0x${device.deviceNetworkId} 1 6 1 {}"
 }
 
 def off() {
-    log.debug "off()"
+    def linkText = getLinkText(device)
+    log.debug "close ${linkText}"
+
+    // only change the state if the vent is not obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot close because ${linkText} is obstructed")
+        return
+    }
+
     sendEvent(makeOnOffResult(0))
     "st cmd 0x${device.deviceNetworkId} 1 6 0 {}"
 }
 
-// does this work?
-def toggle() {
-    log.debug "toggle()"
+def clearObstruction() {
+    def linkText = getLinkText(device)
+    log.debug "attempting to clear ${linkText} obstruction"
 
-    "st cmd 0x${device.deviceNetworkId} 1 6 2 {}"
+    sendEvent([
+        name: "switch",
+        value: "clearing",
+        descriptionText: "${linkText} is clearing obstruction"
+    ])
+
+    // send a move command to ensure level attribute gets reset for old, buggy firmware
+    // then send a reset to factory defaults
+    // finally re-configure to ensure reports and binding is still properly set after the rtfd
+    [
+        makeLevelCommand(device.currentValue("level")), "delay 500",
+        "st cmd 0x${device.deviceNetworkId} 1 0 0 {}", "delay 5000"
```

Comment:
Use `zigbee.zigbeeCommand(0, 0, "{}")` instead. The signature is `zigbeeCommand(cluster, attributeId, values)`